File: training-data/julia/lru_cache.jl
mutable struct LRUCache
    capacity::Int
    map::Dict{Int, Int}
    order::Vector{Int}
    LRUCache(cap::Int) = new(cap, Dict{Int, Int}(), Int[])
end

function get_value(cache::LRUCache, key::Int)
    if haskey(cache.map, key)
        filter!(k -> k != key, cache.order)
        push!(cache.order, key)
        return cache.map[key]
    end
    return nothing
end

function put!(cache::LRUCache, key::Int, value::Int)
    if haskey(cache.map, key)
        filter!(k -> k != key, cache.order)
    end
    cache.map[key] = value
    push!(cache.order, key)
    if length(cache.order) > cache.capacity
        evict = popfirst!(cache.order)
        delete!(cache.map, evict)
    end
end

cache = LRUCache(2)
put!(cache, 1, 1)
put!(cache, 2, 2)
println(get_value(cache, 1))
put!(cache, 3, 3)
println(get_value(cache, 2))

File: training-data/haskell/LRUCache.hs
import Data.IORef
import qualified Data.Map as Map
import Data.Map (Map)

data LRU = LRU { capacity :: Int, store :: Map Int Int, order :: [Int] }

newLRU :: Int -> IO (IORef LRU)
newLRU cap = newIORef (LRU cap Map.empty [])

touch :: Int -> [Int] -> [Int]
touch k xs = k : filter (/= k) xs

get :: IORef LRU -> Int -> IO (Maybe Int)
get ref k = do
  lru <- readIORef ref
  case Map.lookup k (store lru) of
    Nothing -> return Nothing
    Just v -> do
      writeIORef ref lru { order = touch k (order lru) }
      return (Just v)

put :: IORef LRU -> Int -> Int -> IO ()
put ref k v = do
  lru <- readIORef ref
  let newStore = Map.insert k v (store lru)
      newOrder = touch k (order lru)
  if Map.size newStore > capacity lru
    then do
      let evict = last newOrder
      writeIORef ref lru { store = Map.delete evict newStore, order = init newOrder }
    else writeIORef ref lru { store = newStore, order = newOrder }

main :: IO ()
main = do
  ref <- newLRU 2
  put ref 1 1
  put ref 2 2
  v1 <- get ref 1
  put ref 3 3
  v2 <- get ref 2
  print (v1, v2)
